File: backend/app/schemas/employee.py
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List
from datetime import datetime

class EmployeeBase(BaseModel):
    last_name: str = Field(..., min_length=1, max_length=20)
    first_name: str = Field(..., min_length=1, max_length=10)
    title: Optional[str] = Field(None, max_length=30)
    title_of_courtesy: Optional[str] = Field(None, max_length=25)
    birth_date: Optional[datetime] = None
    hire_date: Optional[datetime] = None
    address: Optional[str] = Field(None, max_length=60)
    city: Optional[str] = Field(None, max_length=15)
    region: Optional[str] = Field(None, max_length=15)
    postal_code: Optional[str] = Field(None, max_length=10)
    country: Optional[str] = Field(None, max_length=15)
    home_phone: Optional[str] = Field(None, max_length=24)
    extension: Optional[str] = Field(None, max_length=4)
    notes: Optional[str] = None
    reports_to: Optional[int] = None
    photo_path: Optional[str] = Field(None, max_length=255)

class EmployeeCreate(EmployeeBase):
    pass

class EmployeeUpdate(BaseModel):
    last_name: Optional[str] = Field(None, min_length=1, max_length=20)
    first_name: Optional[str] = Field(None, min_length=1, max_length=10)
    title: Optional[str] = Field(None, max_length=30)
    title_of_courtesy: Optional[str] = Field(None, max_length=25)
    birth_date: Optional[datetime] = None
    hire_date: Optional[datetime] = None
    address: Optional[str] = Field(None, max_length=60)
    city: Optional[str] = Field(None, max_length=15)
    region: Optional[str] = Field(None, max_length=15)
    postal_code: Optional[str] = Field(None, max_length=10)
    country: Optional[str] = Field(None, max_length=15)
    home_phone: Optional[str] = Field(None, max_length=24)
    extension: Optional[str] = Field(None, max_length=4)
    notes: Optional[str] = None
    reports_to: Optional[int] = None
    photo_path: Optional[str] = Field(None, max_length=255)

class EmployeeResponse(EmployeeBase):
    employee_id: int
    created_at: datetime
    updated_at: datetime
    reports_to_name: Optional[str] = None
    
    model_config = ConfigDict(from_attributes=True)

class EmployeeListResponse(BaseModel):
    employee_id: int
    last_name: str
    first_name: str
    title: Optional[str]
    city: Optional[str]
    country: Optional[str]
    reports_to_name: Optional[str] = None
    
    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/employee.ts
export interface EmployeeStatistics {
    total_orders: number;
    orders_this_month: number;
    total_sales: number;
    average_order_value: number;
}

export interface ManagerInfo {
    employee_id: number;
    full_name: string;
    title: string | null;
}

export interface SubordinateInfo {
    employee_id: number;
    full_name: string;
    title: string | null;
}

export interface Employee {
    employee_id: number;
    last_name: string;
    first_name: string;
    title: string | null;
    title_of_courtesy: string | null;
    birth_date: string | null;
    hire_date: string | null;
    address: string | null;
    city: string | null;
    region: string | null;
    postal_code: string | null;
    country: string | null;
    home_phone: string | null;
    extension: string | null;
    notes: string | null;
    reports_to: number | null;
    photo_path: string | null;
    reports_to_name?: string | null;

    // Extended info
    manager?: ManagerInfo;
    subordinates?: SubordinateInfo[];
    statistics?: EmployeeStatistics;

    // Computed/Frontend helpers
    full_name?: string; // Often comes from backend or computed
}

export interface EmployeeFormData {
    last_name: string;
    first_name: string;
    title?: string;
    title_of_courtesy?: string;
    birth_date?: string | null;
    hire_date?: string | null;
    address?: string;
    city?: string;
    region?: string;
    postal_code?: string;
    country?: string;
    home_phone?: string;
    extension?: string;
    notes?: string;
    reports_to?: number | null;
    photo_path?: string;
}

export interface OrgNode {
    employee_id: number;
    name: string;
    title: string | null;
    subordinates: OrgNode[];
}

import { PaginationParams } from './api';

export interface EmployeeFilterParams extends PaginationParams {
    search?: string;
    title?: string;
    city?: string;
    country?: string;
    sort_by?: string;
    sort_order?: 'asc' | 'desc';
}
